### ISSUE TITLE
Text Generation without Webui

### ISSUE THREAD

**Issue by MuhammadShifa** [oobabooga/text-generation-webui#2373]
I want to run it without Gradio. I just need to run it from command terminal having user_input: and model output: , if anyone has script for that kindly share.

**Comment by thusinh1969** [comment]
You can call API, pretty simple !

**Comment by MuhammadShifa** [comment]
Thanks for your reply @thusinh1969 , Actually I need a simple flask-based application that I can integrate with my further customized module.

**Comment by AntouanK** [comment]
@thusinh1969 is there documentation for it?

**Comment by ElevenGameStudios** [comment]
there isn't a lot of api documentation, but just look at /extensions/api/ folder. [blocking_api.py](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/api/blocking_api.py) implements a REST API, while [streaming_api.py](https://github.com/oobabooga/text-generation-webui/blob/main/extensions/api/streaming_api.py) uses websockets.

**Comment by ElevenGameStudios** [comment]
And of course you have to start oobabooga with the --api flag. The api scripts are pretty much self explanatory, you can see all the command uris and all inputs and outputs are UTF8 json format...

**Comment by AntouanK** [comment]
thanks.
I'll try it.
At the moment I have this issue https://github.com/oobabooga/text-generation-webui/issues/2384 so I cannot do anything really.

The --api goes in webui.py, right?

**Comment by AntouanK** [comment]
got it working now. thanks.

how can you switch to chat/chat-instruct/instruct mode? ( not sure what the actual difference but I always select chat in the UI )
Also, how to select what character/persona you are talking to?
( from the API I mean )

**Comment by ElevenGameStudios** [comment]
The modes don't really exist at API level, because they just determine how the input prompt is formatted. You can turn Oobabooga to verbose logging mode to see the exact prompts for the different modes. Same goes for persona, it is just part of the input prompt, if you turn on verbose, you will see how to construct such a prompt. good luck!

**Comment by AntouanK** [comment]
So when I make a call to /generate, does it use whatever character I have selected in the UI?
Or not?
If not, I can prepend the context to the prompt in every API call I guess, I can control it that way.

About chat/chat-instruct/instruct, I still don't understand the difference to be honest.
But what I guess is the issue is the context in the API calls.
I suppose I can do the same, prepend the previous prompt-response to the new prompt I want to send?

I'm fine to do all that, just trying to understand what the /generate call does already, to make sure I can reason the responses I get.

( also, the /chat endpoint I cannot get it to work. does anyone have any sample JSON that works and explain the .history part? )

Thanks for all the help 👍

**Comment by AntouanK** [comment]
for example : 

<img width="769" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/4569111/47d0988b-98da-48f7-8d10-2c81a98a8000">

and without context

<img width="764" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/4569111/5f7f66fa-283c-43b9-8e5d-1f6790729b58">


and when I try something different

<img width="1401" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/4569111/9c218a6d-4eac-49c4-b5b5-03f50422aa20">


So yeah, I don't get how to set the character/context really.

**Comment by AntouanK** [comment]
@ElevenGameStudios 
ah, that `--verbose` flag did help.

```

### Instruction:
Continue the chat dialogue below. Write a single reply for the character "yesorno".

only reply with yes or no! no other answer is acceptable!
You: Hello there. Who are you?


### Response:
yesorno:
--------------------

Output generated in 1.67 seconds (5.40 tokens/s, 9 tokens, context 61, seed 1094402828)```

I think I got it now. It's kind of freestyle though, right? There's no certain formatting that works best, correct?

**Comment by ElevenGameStudios** [comment]
Yes. Also try and see how it looks with some more dialogue history. Generally speaking the input prompt is contrusted like that: Character Persona + Input History + Current Prompt. Overall one has to keep track of the token counts for all of those to limit the Input History so it stays below the token limit.

**Comment by AntouanK** [comment]
Thank you.
On the token count, should I expect that the prompt I'll give ( I mean the sum of context + "prompt" ) plus the answer, will always be maximum total of 2048 tokens?

In the UI there's this section
<img width="1149" alt="image" src="https://github.com/oobabooga/text-generation-webui/assets/4569111/023aafe7-33dc-41bf-8827-efd5c96884fe">
I didn't understand why the "max_new_tokens" is 200.
If the total is always 2048, just take the prompt and use the rest. Why limit it to 200? what's the reasoning?

( also, can we enable the "Stop generating at new line character" via the API maybe? )

**Comment by ElevenGameStudios** [comment]
yeah, I think you understood the overall token limit right. About the max_new_tokens, I am not entirely sure.. Your question makes sense, but maybe it is limited so you get a more consistent dialogue. If one would always max it out, you'd get very long responses at first, but as soon as the conversation includes some history (and probably most of the time) it would be down to the limit anyway.
You can see the json structure of the prompt request here https://github.com/oobabooga/text-generation-webui/blob/main/api-example.py  One could change the generation settings completely with every single prompt, so you can experiment. Not sure which variable is "Stop generating at new line character", but it should be possible.
In my Unity integration I coded some custom stopping behaviours, that don't really stop the generation process, but can still cut off the prompt results and therefore also what's included into dialogue history. The custom "stopping_strings" didn't always work for me...

**Comment by jaredtalon88** [comment]
> 

I am also unable to get the /api/v1/chat endpoint to work properly. I just get back

```json
{
    "results": [
        {
            "history": {
                "internal": [],
                "visible": []
            }
        }
    ]
}
```

whenever I try by following the example
